File: apps/android/app/src/main/java/org/peerscholar/app/data/LearnerSync.kt
package org.peerscholar.app.data

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await

/**
 * Firestore access for a learner's own progress document, `learnerState/{uid}`
 * (owner-only per firebase/firestore.rules).
 *
 * Matches apps/web/lib/learnerState.ts and the iOS LearnerSync so all three
 * clients read and write the same shape. Every call throws rather than
 * silently succeeding when Firebase isn't wired up, and callers treat failure
 * as "stay local".
 */
object LearnerSync {
    data class State(
        val enrollments: List<String> = emptyList(),
        val completedLessons: Map<String, List<String>> = emptyMap(),
        val bookings: List<String> = emptyList(),
    )

    fun currentUid(): String? = runCatching { FirebaseAuth.getInstance().currentUser?.uid }.getOrNull()

    suspend fun fetch(uid: String): State {
        val snapshot = FirebaseFirestore.getInstance()
            .collection("learnerState")
            .document(uid)
            .get()
            .await()

        if (!snapshot.exists()) return State()

        @Suppress("UNCHECKED_CAST")
        val lessons = (snapshot.get("completedLessons") as? Map<String, List<String>>) ?: emptyMap()

        return State(
            enrollments = (snapshot.get("enrollments") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
            completedLessons = lessons,
            bookings = (snapshot.get("bookings") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
        )
    }

    suspend fun save(uid: String, state: State) {
        FirebaseFirestore.getInstance()
            .collection("learnerState")
            .document(uid)
            .set(
                mapOf(
                    "enrollments" to state.enrollments,
                    "completedLessons" to state.completedLessons,
                    "bookings" to state.bookings,
                    "updatedAt" to System.currentTimeMillis(),
                ),
                com.google.firebase.firestore.SetOptions.merge(),
            )
            .await()
    }
}

File: apps/ios/PeerScholar/ViewModels/LearnerSync.swift
import Foundation
import FirebaseCore
import FirebaseAuth
import FirebaseFirestore

/// Firestore access for a learner's own progress document, `learnerState/{uid}`
/// (owner-only per firebase/firestore.rules).
///
/// Matches the web app's apps/web/lib/learnerState.ts so both clients read and
/// write the same shape. Every call no-ops safely when Firebase isn't
/// configured, so the app still runs on mock data with no project attached.
enum LearnerSync {
    struct State {
        var enrollments: [String]
        var completedLessons: [String: [String]]
        var bookings: [String]

        static let empty = State(enrollments: [], completedLessons: [:], bookings: [])
    }

    enum SyncError: Error {
        case notConfigured
    }

    static var isConfigured: Bool { FirebaseApp.app() != nil }

    static var currentUid: String? {
        guard isConfigured else { return nil }
        return Auth.auth().currentUser?.uid
    }

    static func fetch(uid: String) async throws -> State {
        guard isConfigured else { throw SyncError.notConfigured }
        let snapshot = try await Firestore.firestore().collection("learnerState").document(uid).getDocument()
        guard let data = snapshot.data() else { return .empty }
        return State(
            enrollments: data["enrollments"] as? [String] ?? [],
            completedLessons: data["completedLessons"] as? [String: [String]] ?? [:],
            bookings: data["bookings"] as? [String] ?? []
        )
    }

    static func save(uid: String, state: State) async throws {
        guard isConfigured else { throw SyncError.notConfigured }
        try await Firestore.firestore().collection("learnerState").document(uid).setData(
            [
                "enrollments": state.enrollments,
                "completedLessons": state.completedLessons,
                "bookings": state.bookings,
                "updatedAt": ISO8601DateFormatter().string(from: Date()),
            ],
            merge: true
        )
    }
}
